covariance/covarianceBase.cpp

```diff
@@ -1393,125 +1389,266 @@ void covarianceBase::updateThrowMatrix(TMatrixDSym *cov){
+
+    // Should check this is done
+    if(means_vector->GetNrows()){
+      MACH3LOG_ERROR("External means vec size ({}) != matrix size ({})", means_vector->GetNrows(), size);
+      throw;
+    }
+
+    par_means = std::vector<double>(size);
+    for(int i=0; i<size; i++){
+      par_means[i] = (*means_vector)(i);
+    }
+
+  }
+  // Totally fine if it doesn't exist, we just can't do adaption
+  else{
+    // We don't need a means vector, set the adaption=false
+    MACH3LOG_WARN("Cannot find means vector in {}, therefore I will not be able to adapt!", matrix_file_name);
+    use_adaptive=false;
   }
-  delete tmp_throwMatrix;
-  delete tmp_meansvec;
-  throwMatrixFile->Close();
-  delete throwMatrixFile; // Just in case
+
+
+  matrix_file->Close();
+
+  std::cout<<"Set up matrix from external file"<<std::endl;
+
 }
 
+void covarianceBase::setAdaptiveBlocks(std::vector<std::vector<int>> block_indices){
+  /*
+    In order to adapt efficient we want to setup our throw matrix to be a serious of block-diagonal (ish) matrices
+
+    To do this we set sub-block in the config by parameter index. For example having
+    [[0,4],[4, 6]] in your config will set up two blocks one with all indices 0<=i<4 and the other with 4<=i<6
+  */
+  // Set up block regions
+  adapt_block_matrix_indices = std::vector<int>(getNpars(), 0);
 
-void covarianceBase::useSeparateThrowMatrix(){
-    initialiseNewAdaptiveChain(); 
-    updateThrowMatrix(covMatrix);
+  // Should also make a matrix of block sizes
+  adapt_block_sizes = std::vector<int>((int)block_indices.size()+1, 0);
+  adapt_block_sizes[0]=getNpars();
+
+  if(block_indices.size()==0 || block_indices[0].size()==0) return;
+
+
+  // Now we loop over our blocks
+  for(int iblock=0; iblock<(int)block_indices.size(); iblock++){
+    // Loop over blocks in the block
+    for(int isubblock=0; isubblock<(int)block_indices[iblock].size()-1; isubblock+=2){
+      int block_lb = block_indices[iblock][isubblock];
+      int block_ub = block_indices[iblock][isubblock+1];
+
+      //      std::cout<<block_lb<<" "<<block_ub<<std::endl;
+      
+      if(block_lb>getNpars() || block_ub>getNpars()){
+	MACH3LOG_ERROR("Cannot set matrix block with edges {}, {} for matrix of size {}",
+		       block_lb, block_ub, getNpars());
+	throw;
+      }
+
+      for(int ipar=block_lb; ipar<block_ub; ipar++){
+	adapt_block_matrix_indices[ipar]=iblock+1;
+        adapt_block_sizes[iblock+1]+=1;
+        adapt_block_sizes[0]-=1;
+      }
+
+    }
+  }
+  for(auto i : adapt_block_matrix_indices){
+    std::cout<<i<<" ,";
+  }
+  std::cout<<std::endl;
+  std::cout<<adapt_block_matrix_indices.size()<<std::endl;
 }
 
-//HW: Truly adaptive MCMC!
+void covarianceBase::createNewAdaptiveCovariance(){
+  // If we don't have a covariance matrix to start from for adaptive tune we need to make one!
+  use_adaptive=true;
+  adaptiveCovariance = new TMatrixDSym(size);
+  adaptiveCovariance->Zero();
+  par_means = std::vector<double>(size, 0);
```

Comment:
I think I just noticed this but adaptiveCovariance  is not being destroyed in destructor